File: 조준장/level2/택배상자.py
from collections import deque

def solution(order):
    answer = 0
    n = len(order)
    box = deque([i for i in range(1, n+1)])
    order = deque(order)
    stack = deque([])
    
    for _ in range(n*4):
        if not stack and not order and not box:
            break
            
        if order and box and order[0] == box[0]:
            box.popleft()
            order.popleft()
            answer += 1
        elif stack and order and stack[-1] == order[0]:
            order.popleft()
            stack.pop()
            answer += 1
        else:
            if box:
                stack.append(box.popleft())

    return answer

File: 조준장/level2/택배상자.kt
import java.util.*

class Solution {
    fun solution(order: IntArray): Int {
        var answer: Int = 0
        var box = 1
        var index = 0
        val temp = Stack<Int>()

        while (index < order.size) {
            if (order[index] == box) {
                index ++ 
            } else if (order[index] > box) {
                temp.push(box)
            } else {
                if (temp.size == 0) break
                if (temp.pop() == order[index]) {
                    index ++
                    continue
                } else {
                    break
                }
            }
            box ++
        }
        
        
        return index
    }
}
